Replace PyPI with Datadog index for the pip + TUF default use case.

trishankatdatadog opened the pull request:
### What does this PR do?

If the user has not explicitly overrriden the pip indices, then we replace the PyPI index with our own by default, in order to prevent accidental installation of Datadog or even third-party packages from PyPI.

### Motivation

In order to tighten by default any misconfiguration of security.

### Additional Notes

N/A.

trishankatdatadog committed:
Replace PyPI with Datadog index for the pip + TUF default use case.

Signed-off-by: Trishank K Kuppusamy <trishank.kuppusamy@datadoghq.com>

trishankatdatadog committed:
minor grammar edit

Signed-off-by: Trishank K Kuppusamy <trishank.kuppusamy@datadoghq.com>

trishankatdatadog committed:
minor syntax fix

Signed-off-by: Trishank K Kuppusamy <trishank.kuppusamy@datadoghq.com>

trishankatdatadog committed:
let gofmt reformat code

Signed-off-by: Trishank K Kuppusamy <trishank.kuppusamy@datadoghq.com>

trishankatdatadog committed:
add reno

Signed-off-by: Trishank K Kuppusamy <trishank.kuppusamy@datadoghq.com>

zippolyte commented on cmd/agent/app/integrations.go:
```diff
@@ -197,6 +198,12 @@ func tuf(args []string) error {
 	idxFlags, err := tufCmd.Flags().GetStringSlice("idx-flags")
 	if err == nil {
 		args = append(args, idxFlags...)
+	} else {
```
I don't think this should go in the `else` here. An `err != nil` doesn't mean that the user didn't set a pip index

trishankatdatadog replied:
That's true. More like checking for whether idxFlags is empty. Lemme update.

trishankatdatadog committed:
fix conditional assignment

zippolyte commented on cmd/agent/app/integrations.go:
```diff
@@ -196,7 +197,16 @@ func tuf(args []string) error {
+		if len(idxFlags) > 0 {
+			args = append(args, idxFlags...)
+		} else {
+			// Otherwise, if the user has not explicitly overrriden the pip indices,
```
typo: `overriden`

zippolyte commented on releasenotes/notes/tighten-pip-tuf-index-71cc1530c1ebe930.yaml:
```diff
@@ -0,0 +1,14 @@
+---
+security:
+  - |
+    If the user has not explicitly overrriden the pip indices, then we replace
```
ditto

zippolyte commented on releasenotes/notes/tighten-pip-tuf-index-71cc1530c1ebe930.yaml:
```diff
@@ -0,0 +1,14 @@
+    If the user has not explicitly overrriden the pip indices, then we replace
+    the PyPI index with our own by default, in order to prevent accidental
+    installation of Datadog or even third-party packages from PyPI. The user
+    may override this default behaviour using the `--idx-flags` option.
```
Let's not mention that flag since it's not documented

trishankatdatadog committed:
fix typos

trishankatdatadog commented:
Can we change the milestone from 6.7.0 to 6.6.0, please?

zippolyte reviewed (dismissed):
Looks good to me, let's just fix the conflict and have someone from @DataDog/agent-core approve. I think we can have it in 6.6.0

ofek committed:
Merge branch 'master' into trishankatdatadog/tighten-pip-tuf-index

ofek committed:
fix style

trishankatdatadog commented:
Can someone from @DataDog/agent-core approve, please? Thanks! 🙇

gmmeyer reviewed (dismissed):
I'm happy with this!

trishankatdatadog commented:
Thanks for the approval, guys! I will double-check this tomorrow to make sure it works. Let's hold off merging until then.

trishankatdatadog committed:
Correctly set `pip install --index-url`

Also remove some undocumented power user options that were not working
as intended

trishankatdatadog commented:
@masci @ofek @olivielpeau I have:

(1) Correctly set the `--index-url` option for `pip install` and `pip search`.
(2) Removed the undocumented "--cmd-flags" and "--idx-flags" power user options to the `datadog-agent integration` command that did not work as expected.

I would appreciate your quick review and approval so that we can merge this for RC2 today. Thanks! 🙇

zippolyte commented on cmd/agent/app/integrations.go:
```diff
@@ -444,6 +425,10 @@ func freeze(cmd *cobra.Command, args []string) error {
+		// We replace the PyPI index with our own by default, in order to prevent
+		// accidental installation of Datadog or even third-party packages from
+		// PyPI.
+		"--index-url", tufIndex,
```
Is that necessary for `pip freeze` ? I thought it just read the local metadata

trishankatdatadog replied:
Good catch, thanks, it's meant to be for `pip search` --- I'll fix right away!

zippolyte reviewed (dismissed):
LGTM

olivielpeau reviewed (dismissed):
👍

trishankatdatadog committed:
set our index for the `pip search`, not `pip freeze`, cmd

olivielpeau commented on releasenotes/notes/tighten-pip-tuf-index-71cc1530c1ebe930.yaml:
```diff
@@ -0,0 +1,16 @@
+    We replace the PyPI index with our own by default, in order to prevent
+    accidental installation of Datadog or even third-party packages from PyPI.
+  - |
+    Remove some undocumented power user options to the `datadog integration`
```
format of the release note is reST: inline code needs to be quoted with double backticks instead of single backticks :)

also, it's the `datadog-agent integration` command

olivielpeau commented on releasenotes/notes/tighten-pip-tuf-index-71cc1530c1ebe930.yaml:
```diff
@@ -0,0 +1,16 @@
+security:
+  - |
+    We replace the PyPI index with our own by default, in order to prevent
+    accidental installation of Datadog or even third-party packages from PyPI.
```
anything under `security` will be listed as `Security Issues` in the changelog, so we need to be extra-careful with how we word them.

Could you:
* add more context here (in particular mention the `datadog-agent integration` command)
* maybe put these 2 release notes under `fixes` unless they're actual security issues. If we keep them under security issues let's double check all users understand that this isn't a serious security issue affecting their install

olivielpeau reviewed (commented):
Happy with the code changes, just a few notes on the release note

trishankatdatadog committed:
Address @olivielpeau comments

ofek reviewed (approved):
LGTM now too

trishankatdatadog commented:
Thanks for the quick review, guys! @olivielpeau, do the release notes look better now?

olivielpeau reviewed (approved):
👌

trishankatdatadog commented:
Thanks for the quick response---you guys are awesome!
